File: src/main/java/com/tngtech/jgiven/scenario/state/JGivenUsageProvider.java
package com.tngtech.jgiven.scenario.state;

import com.intellij.psi.PsiField;
import com.intellij.psi.PsiReference;
import com.tngtech.jgiven.resolution.ResolutionHandler;

import javax.annotation.Nullable;

class JGivenUsageProvider {

    private final ScenarioStateAnnotationProvider scenarioStateProvider;
    private final ResolutionHandler resolutionHandler;
    private final ReferenceFactory referenceFactory;

    JGivenUsageProvider(ScenarioStateAnnotationProvider scenarioStateProvider, ResolutionHandler resolutionHandler, ReferenceFactory referenceFactory) {
        this.scenarioStateProvider = scenarioStateProvider;
        this.resolutionHandler = resolutionHandler;
        this.referenceFactory = referenceFactory;
    }

    @Nullable
    PsiReference createReferenceIfJGivenUsage(PsiField fieldToSearch, PsiField field) {
        if (scenarioStateProvider.isJGivenScenarioState(field)
                && !fieldToSearch.equals(field)
                && resolutionHandler.resolutionMatches(field, fieldToSearch)) {
            return referenceFactory.referenceFor(field);
        }
        return null;
    }
}

File: src/main/java/com/tngtech/jgiven/scenario/state/JGivenUsageProvider.kt
package com.tngtech.jgiven.scenario.state

import com.intellij.psi.PsiField
import com.tngtech.jgiven.resolution.ResolutionHandler

class JGivenUsageProvider(
        private val scenarioStateProvider: ScenarioStateAnnotationProvider,
        private val resolutionHandler: ResolutionHandler,
        private val referenceFactory: ReferenceFactory
) {
    fun createReferenceIfJGivenUsage(fieldToSearch: PsiField, field: PsiField) =
            when {
                scenarioStateProvider.isJGivenScenarioState(field)
                        && fieldToSearch != field
                        && resolutionHandler.resolutionMatches(field, fieldToSearch) -> referenceFactory.referenceFor(field)
                else -> null
            }
}
